nimbus/src/enrollment.rs

```diff
@@ -0,0 +1,340 @@
+        branch: String,
+    },
+    // Not enrolled because our evaluator declined to choose us.
+    NotSelected,
+    // Not enrolled because either enrollment is paused or the experiment is over.
+    NotRunning,
+    // Not enrolled because we are not being targeted for this experiment.
+    NotTargeted,
+    // User explicitly opted out.
+    OptedOut,
+    // There was some error opting in.
+    Error {
+        // Ideally this would be an Error, but then we'd need to make Error
+        // serde compatible, which isn't trivial nor desirable.
+        reason: String,
+    },
+}
+
+/// Return information about all enrolled experiments.
+pub fn get_enrollments(db: &Database) -> Result<Vec<EnrolledExperiment>> {
+    let enrollments: Vec<ExperimentEnrollment> = db.collect_all(StoreId::Enrollments)?;
+    let mut result = Vec::with_capacity(enrollments.len());
+    for enrollment in enrollments {
+        log::debug!("Have enrollment: {:?}", enrollment);
+        if let EnrollmentStatus::Enrolled { ref branch, .. } = enrollment.status {
+            if let Some(experiment) =
+                db.get::<Experiment>(StoreId::Experiments, &enrollment.slug)?
+            {
+                result.push(EnrolledExperiment {
+                    slug: experiment.slug,
+                    user_facing_name: experiment.user_facing_name,
+                    user_facing_description: experiment.user_facing_description,
+                    branch_slug: branch.to_string(),
+                });
+            } else {
+                log::warn!(
+                    "Have enrollment {:?} but no matching experiment!",
+                    enrollment
+                );
+            }
+        }
+    }
+    Ok(result)
+}
+
+/// Update all enrollments. Typically used immediately after the list of
+/// experiments has been refreshed, so some might mean new enrollments, some
+/// might have expired, etc.
+pub fn update_enrollments(
+    db: &Database,
+    nimbus_id: &Uuid,
+    aru: &AvailableRandomizationUnits,
+    app_context: &AppContext,
+) -> Result<()> {
+    log::info!("updating enrollments...");
+    // We might have enrollments for experiments which no longer exist, so we
+    // first build a set of all IDs in both groups.
+    let mut all_slugs = HashSet::new();
+    let experiments = db.collect_all::<Experiment>(StoreId::Experiments)?;
+    let mut map_experiments = HashMap::with_capacity(experiments.len());
+    for e in experiments {
+        all_slugs.insert(e.slug.clone());
+        map_experiments.insert(e.slug.clone(), e);
+    }
+    // and existing enrollments.
+    let enrollments = db.collect_all::<ExperimentEnrollment>(StoreId::Enrollments)?;
+    let mut map_enrollments = HashMap::with_capacity(enrollments.len());
+    for e in enrollments {
+        all_slugs.insert(e.slug.clone());
+        map_enrollments.insert(e.slug.clone(), e);
+    }
+
+    // XXX - we want to emit events for many of these things, but we don't have
```

Comment:
We could also return a "Diff" structure to the Kotlin layer, and let them emit observable events to the app.